redirectToAuthentication if the request uses PAR to not lose the single object after a refresh

rmartinc opened the pull request:
Closes #36716

The [rfc9126 section 4](https://datatracker.ietf.org/doc/html/rfc9126#section-4) states that:

>  Authorization servers SHOULD treat request_uri values as one-time use but MAY allow for duplicate requests due to a user reloading/refreshing their user agent.

So using a single-use object for the PAR data is OK but I think that it should survive for refreshes of the page. This is specially problematic for spnego login because there is always a second request triggered by the 401 error asking for the kerberos token.

This PR just stores the PAR data into the auth session (client notes). This way the data continues to be a single use object but it remains for the browser session that initially recovered it. I also managed the idea of re-adding the object in the single-use store but I think this is better because we are never sure how many times the page is going to be refreshed.

Tests added.

dteleguin commented:
@rmartinc thanks for your work!

I would _strongly_ suggest that we make this behavior configurable and disabled by default. Single-use PAR has been in Keycloak for years (literally, since 2021). The existing systems might be relying on PAR `request_uri`s being one-time, and changing this might introduce non-obvious security vulnerabilities.

rmartinc commented:
@dteleguin With the PR the PAR data is still a one-time object but once consumed by an auth request it is maintained in that browser until the final login or expiration. But yes, this can have unwanted collateral effects like any change in general. Another thing we can try to do it is just doing this if kerberos is detected (not sure if possible, but maybe we can try it). With that idea idea (or similar) normal reloads will make the PAR request fail, but this is already happening right now, so I don't see it as a problem.

rmartinc committed:
redirectToAuthentication if the request uses PAR to not lose the single object after a refresh
Closes #36716

Signed-off-by: rmartinc <rmartinc@redhat.com>

rmartinc commented:
Thanks a lot @mposolda! Yes, much easier this way, and we do the same, the PAR request is retrieved from the single use store and saved in the auth session. As the authentication is managed later in a subsequent request, the same auth session is used and everything works, even reloading the page. The PR is very small now.

I have also removed the changes in the kerberos test. Now the ParTest perform two refreshes to ensure the request survives.

Rebased too as this PR was a bit old. Let's see if the CI passes completely.

mposolda reviewed (approved):
@rmartinc Cool, Thanks!
